more_autograding_examples/jupyter_notebook_autograding/config/test_input/grader.py

```diff
@@ -0,0 +1,78 @@
+from pathlib import Path
+import nbformat
+from nbconvert.preprocessors import ClearOutputPreprocessor, ExecutePreprocessor, CellExecutionError
+import base64
+
+# Execution API: https://nbconvert.readthedocs.io/en/latest/execute_api.html
+def execute_notebook(timeout=600):
+    
+    notebook_files = list(Path('.').glob('*.ipynb'))
+    if not notebook_files:
+        raise FileNotFoundError("No Jupyter notebook files found in the current directory.")
+    notebook_filename = Path(notebook_files[0])
+
+    with open(notebook_filename) as f:
+        # Load the notebook and do not convert it to a specific version
+        nb = nbformat.read(f, as_version=nbformat.NO_CONVERT)
+
+    # Clear output of notebook
+    cop = ClearOutputPreprocessor()
+    nb, _ = cop.preprocess(nb, {})
+
+    # Sets preprocessor with the following parameters:
+    # - timeout: maximum time in seconds to execute a cell
+    # - kernel_name: the kernel to use for execution (e.g., 'python3')
+    # - allow_errors: if True, allows the execution to continue even if a cell raises an error (Optional, defaults to False)
+    ep = ExecutePreprocessor(timeout=timeout, kernel_name='python3', allow_errors=False)
+
+    try:
+        out = ep.preprocess(nb, {'metadata': {'path': notebook_filename.parent}})
+    # CellExecutionError is raised if a cell execution fails and allow_errors is False
+    except CellExecutionError:
+        out = None
+        msg = 'Error executing the notebook "%s".\n\n' % notebook_filename
+        msg += 'See notebook executed.ipynb for the traceback.'
+        print(msg)
+        raise
+    
+    with open("executed.ipynb", mode='w', encoding='utf-8') as f:  
+        nbformat.write(nb, f)
+    
+    cell_idx = 0
```

Comment:
I think we should probably run the index/counter for the cells in the notebook for ALL cells, not just the code cells.  If the student is writing markdown (e.g., as documention, or written/non-code answer) as part of their assignment, the instructor may wish to extract and process or grade these cells by index/id as well.